File: python/interview_36.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if not root:
            return None
        self.preNode = None
        self.head = None

        def inorderVisit(node):
            if not node:
                return
            inorderVisit(node.left)
            if self.preNode:
                self.preNode.right = node
                node.left = self.preNode
            else:
                self.head = node
            self.preNode = node
            inorderVisit(node.right)

        inorderVisit(root)
        self.preNode.right = self.head
        self.head.left = self.preNode
        return self.head

File: Java/interview_36.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,Node _left,Node _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/
class Solution {
    private Node head;
    private Node preNode;

    private void inorderVisit(Node node) {
        if (node == null) {
            return;
        }
        inorderVisit(node.left);
        if (preNode == null) {
            head = node;
        }
        else {
            preNode.right = node;
            node.left = preNode;
        }
        preNode = node;
        inorderVisit(node.right);
    }

    public Node treeToDoublyList(Node root) {
        if (root == null) {
            return null;
        }
        inorderVisit(root);
        head.left = preNode;
        preNode.right = head;
        return head;
    }
}
